### PR TITLE
Alternative screen use

### PR DESCRIPTION
**Pull request by yb66**
Hi,

This would be helpful to me so I tried implementing it, seems to work (I'm using iTerm2 and tried it with Terminal too). I put an example into the final sample, not sure if it's the right place or whether it would need some extra comments or a bit in the readme… Let me know your thoughts.

Regards,
iain

### CODE REVIEW

File: samples/tutorial/step6.cr [hunk 1]
```diff
@@ -89,12 +89,14 @@ if File.exists? HISTFILE # Does it exist?
 end
 
 begin # Get rid of stacktrace on ^C
+	fancy.tty.switch_to_alternate_screen
```
Comment (Papierkorb) [inline samples/tutorial/step6.cr]: The indention is off, please check if your editor is set to use two spaces per level.  Most added lines suffer from this.

File: src/fancyline/tty/vt100.cr [hunk 2]
```diff
@@ -90,6 +90,14 @@ class Fancyline
         @io.print "\e[2J\e[H"
       end
 
+			def switch_to_alternate_screen
```
Comment (Papierkorb) [inline src/fancyline/tty/vt100.cr:93]: The methods you added should also be part of the Tty class as `abstract def`s.  There, document the methods as appropriate.  (Rule of thumb: Every public API should have full docs coverage)

**Review by Papierkorb — CHANGES_REQUESTED** [review]
Hello,

Now that's a cool feature I didn't know about!  I do have some change requests..

Also, I think a yielding method would be good, ala `Tty#in_alternate_screen`.  The method then calls the switch_to method, yields, and upon return switches back again in an rescue-else block.  Ideally, whatever the block returned should be passed back to the caller.  With this less can go wrong on the users end :)

Thank You!

File: src/fancyline/tty/vt100.cr (hunk 2 above)
Comment (yb66) [reply src/fancyline/tty/vt100.cr:93]: Didn't notice that class 😅 No probs, I'll add the docs once we're settled on the API.

File: samples/tutorial/step6.cr (hunk 1 above)
Comment (yb66) [reply samples/tutorial/step6.cr]: My mistake, I prefer tabs but I've set up a commit hook to remove them in future.

**Comment by yb66** [comment]
> Now that's a cool feature I didn't know about!

I only learnt about it yesterday from reading [this code](https://github.com/dylanaraps/fff/blob/master/fff#L31). Seemed like a nice thing to try as I'm trying to wangle Fancyline into running a TUI. The codes are listed in [this helpful gist](https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#common-private-modes) with some others that may come in handy.

>  I think a yielding method would be good, ala Tty#in_alternate_screen. The method then calls the switch_to method, yields, and upon return switches back again in an rescue-else block.

I woke up and found myself thinking the same (great minds:). Not sure what I'd be rescuing though as I think most terminals (from my research yesterday) will support this or have a setting to ignore. I'm going to track down a test that's better than me manually checking.

Something like this?

```crystal
def in_alternate_screen(&block)
  switch_to_alternate_screen
rescue # something
else
  yield
ensure
  switch_from_alternate_screen
end
```

I tried it (also with the switch back in the rescue) but they both failed to switch back. An ensure works:

```crystal
def in_alternate_screen(&block)
  switch_to_alternate_screen
  yield
ensure
  switch_from_alternate_screen
end
```

Let me know if this is along the lines you were thinking.

**Comment by yb66** [comment]
I also wonder about adding a convenience wrapper in Fancyline to hide the tty part from the consumer. I'm not sure they should care that the tty is handling the screens, e.g.

    fancy.tty.in_alternate_screen do

could become

    fancy.in_alternate_screen do

It's not much but keeping the consumer of the library at the top level of the API is nice. Just an idea.

**Comment by Papierkorb** [comment]
> An ensure works

Ah sorry, that's what I meant 👍 in the gist you linked it says 

> Note: While these modes may be supported by the most terminals, some may not work in multiplexers like tmux.

Could you check what happens in screen and tmux? Do they notify us that the operation failed somehow?

**Comment by lagerfeuer** [comment]
tmux is a big one for me, I don't use a terminal without it. @yb66 let me know if I can assist you somehow.

Also, I added a CI style check, to give you some feedback if the coding style is inconsistent. A `crystal tool format` commit hook is probably the way to go.

**Comment by yb66** [comment]
Sorry to leave this hanging, had an urgent matter come up and was constantly thinking “tomorrow, tomorrow...” but tomorrow never arrived!

I’ve some free time coming up so I’ll get to this asap.

Regards,
iain

**Comment by lagerfeuer** [comment]
Hey, don't worry about it, you'll get to it when you get to it 😄 
I'm just happy people wanna help out!

**Comment by lagerfeuer** [comment]
Stale PR, closing.

**Pull request closed without merging**